### PR TITLE
Add Asm.js React Native build

### PR DESCRIPTION
**Pull request by Sekhmet**
Files:
- `monero_utils/MyMoneroCoreCpp_ASMJS.asm.rn.js`
- `monero_utils/MyMoneroCoreCpp_ASMJS.rn.js`

are generated using `bin/package-asm-rn.sh` script.

`index.rn.js` is `monero_utils/MyMoneroCoreBridge.js` without Node.js and Web-specific APIs (`readFileSync`/`fetch`).

### CODE REVIEW

File: bin/package-asm-rn.sh [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+LOADER_FILE="$DIR/MyMoneroCoreCpp_ASMJS.js"
+LOADER_FILE_OUT="$DIR/MyMoneroCoreCpp_ASMJS.rn.js"
+
+echo "🔁  Creating Base64 module for ASMJS"
+echo "module.exports = \"`base64 $ASM_FILE`\";" > $ASM_FILE_OUT
+
+echo "🔁  Patching loader for React Native support"
+sed 's/require("/requirexxx("/g' $LOADER_FILE > $LOADER_FILE_OUT
+sed -i '.bak' 's/requirexxx("crypto")/require("crypto")/g' $LOADER_FILE_OUT
```
Comment (Sekhmet) [inline bin/package-asm-rn.sh:16]: This prevents Metro (React Native bundler) from loading Node-specific modules.

File: package.json [hunk 2]
```diff
@@ -3,6 +3,7 @@
   "version": "1.1.4",
   "description": "The JS library containing the Monero crypto plus lightweight wallet functions behind the official MyMonero apps",
   "main": "index.js",
+  "react-native": "index.rn.js",
```
Comment (Sekhmet) [inline package.json:6]: This only adds a new entry for React Native. Other targets continue using existing code.

**Review by gutenye — APPROVED** [review]
utACK

File: index.rn.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+const { atob } = require("./utils");
+
+function initMonero() {
+	return new Promise((resolve, reject) => {
+		const Module_template = {};
+		let Module = {};
+
+		const content = atob(ASM);
+		try {
+			eval(content);
```
Comment (ChALkeR) [inline index.rn.js:12]: Does that use the local context? Perhaps, passing in just `Module` would be fine?
E.g. `new Function('Module', content)(Module)` (if other local vars are not used).

That would still be an eval, but without the rest of the local vars.

File: utils.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	if (str.length % 4 == 1) {
+		throw new Error(
+			"'atob' failed: The string to be decoded is not correctly encoded.",
+		);
+	}
+	for (
+		let bc = 0, bs = 0, buffer, i = 0;
+		(buffer = str.charAt(i++));
+		~buffer && ((bs = bc % 4 ? bs * 64 + buffer : buffer), bc++ % 4)
+			? (output += String.fromCharCode(255 & (bs >> ((-2 * bc) & 6))))
```
Comment (ChALkeR) [inline utils.js:16]: This blows up the memory and consumes about 100 MiB. Also, this is very slow, i.e. could block for a second, depending on the device.
It's better to not build strings using char-by-char concat.

File: index.rn.js (hunk 3 above)
Comment (Sekhmet) [reply index.rn.js:12]: `eval` is used in the original bridge, but it seems that `new Function('Module', content)(Module)` should work just fine. I will give it a go.
Thanks!

File: utils.js (hunk 4 above)
Comment (Sekhmet) [reply utils.js:16]: You are right. We can replace it with `Buffer.from(input, 'base64').toString('binary')`.

Comment (sonaye) [reply utils.js:16]: We need this change for mobile for sure

File: index.rn.js (hunk 3 above)
Comment (ChALkeR) [reply index.rn.js:12]: Why can't this be just a require in the source form, without base64 encoding, btw?

Comment (Sekhmet) [reply index.rn.js:12]: Metro bundler can't handle it and crashes.

File: utils.js (hunk 4 above)
Comment (sonaye) [reply utils.js:16]: @Sekhmet Are you planning to make this change? not sure if I should wait or continue with brining the Monero PRs to master then make this change

Comment (Sekhmet) [reply utils.js:16]: Sorry, I didn't know you are waiting for those changes before merging PRs. I can do it now.

Comment (sonaye) [reply utils.js:16]: Please do, will audit your mobile PR tonight

**Pull request merged**